src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java

```diff
@@ -0,0 +1,125 @@
+  // If a call to "hash()" is made and less than this much data has been
+  // written, a single JNI call will be made that initializes, hashes, and
+  // cleans up the hasher, rather than making separate calls for each operation.
+  public static final int ONESHOT_THRESHOLD = 8 * 1024;
+  private ByteBuffer buffer = ByteBuffer.allocate(ONESHOT_THRESHOLD);
+  private long hasher = -1;
+
+  public Blake3MessageDigest() {
+    super("BLAKE3");
+  }
+
+  private void flush() {
+    if (hasher == -1) {
+      hasher = Blake3JNI.allocate_and_initialize_hasher();
+    }
+
+    if (buffer.position() > 0) {
+      Blake3JNI.blake3_hasher_update(hasher, buffer.array(), buffer.position());
+      buffer.clear();
+    }
+  }
+
+  public void engineUpdate(byte[] data, int offset, int length) {
+    while (length > 0) {
+      int numToCopy = Math.min(length, buffer.remaining());
+      buffer.put(data, offset, numToCopy);
+      length -= numToCopy;
+      offset += numToCopy;
+
+      if (buffer.remaining() == 0) {
+        flush();
+      }
+    }
+  }
+
+  public void engineUpdate(byte[] data) {
+    engineUpdate(data, 0, data.length);
+  }
+
+  public void engineUpdate(byte b) {
+    if (buffer.remaining() == 0) {
+      flush();
+    }
+    buffer.put(b);
+  }
+
+  private byte[] getOutput(int outputLength) {
+    byte[] retByteArray = new byte[outputLength];
+
+    if (hasher == -1) {
+      // If no flush has happened yet; oneshot this.
+      Blake3JNI.oneshot(buffer.array(), buffer.position(), retByteArray, outputLength);
+    } else {
+      flush();
+      Blake3JNI.blake3_hasher_finalize_and_reset(hasher, retByteArray, outputLength);
+    }
+
+    buffer.clear();
+    return retByteArray;
+  }
+
+  public Object clone() throws CloneNotSupportedException {
+    throw new CloneNotSupportedException();
+  }
+
+  public void engineReset() {
+    if (hasher != -1) {
+      Blake3JNI.blake3_hasher_reset(hasher);
+    }
+    buffer.clear();
+  }
+
+  public void engineUpdate(ByteBuffer input) {
+    if (input.hasArray()) {
+      engineUpdate(input.array());
+    } else {
+      byte[] bufCopy = new byte[input.position()];
+      input.get(bufCopy);
+      engineUpdate(bufCopy);
+    }
+  }
+
+  public int engineGetDigestLength() {
+    return OUT_LEN;
+  }
+
+  public byte[] engineDigest() {
+    byte[] digestBytes = getOutput(OUT_LEN);
+    return digestBytes;
+  }
+
+  public int engineDigest(byte[] buf, int off, int len) throws DigestException {
+    if (len < OUT_LEN) {
+      throw new DigestException("partial digests not returned");
+    }
+    if (buf.length - off < OUT_LEN) {
+      throw new DigestException("insufficient space in the output buffer to store the digest");
+    }
+
+    byte[] digestBytes = getOutput(OUT_LEN);
+    System.arraycopy(digestBytes, 0, buf, off, digestBytes.length);
+    return digestBytes.length;
+  }
+
+  @Override
+  protected void finalize() throws Throwable {
```

Comment:
@coeuvre  Your suggestion is somewhat complicated: is there a precedent / do you have an example? I was unable to find one.

We don't want to pass a byte[] (hasher) for every JNI call instead of the pointer; moving bytes across JNI is not free. So when initializing the hasher we'd need to make one call to JNI that returns a byte[], and then another JNI call to get the address of that object. Do you see a better way to do this? Does Java guarantee that the address of the object on the java heap will remain valid?

I also notice there is precedent for using finalize() which is why I thought this would be alright for now.

lmk what you think!

Reply:
The modern replacement for `finalize` is [`Cleaner`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/ref/Cleaner.html), which I think would be a good fit here.

The address of a `byte` array would not remain fixed. The `GetPrimitiveArrayCritical` JNI method could be used to limit copies, but I don't know whether a hash function runs quickly enough that this wouldn't hurt overall GC performance.

Reply:
Thanks @fmeum -- this is the approach I implemented.

Reply:
@fmeum Thanks for the pointer! I think in this case it is fast enough (we do this internally for our hash function).

@tylerwilliams Thanks for updating the code! Now we use `close()` to free the memory but I think we never call it (please correct me if I am wrong). I thought the reason why we used `finalize` is to avoid manual memory management. On the other hand,  if we use `close()`, we could just free memory there instead of using `Cleaner`.

> Do you see a better way to do this? Does Java guarantee that the address of the object on the java heap will remain valid?

What we do internally is something like:

```
// In Java
class Blake3MessageDigest {
    // The value of `sizeof(blake3_hasher)`. If it doesn't have a well known size, we could query using JNI inside following static block.
    private static int STATE_SIZE = xxx;

    // Cached initial state for a new hasher
    private static byte[] INITIAL_STATE = new byte[STATE_SIZE];

    static {
        nativeInit(INITIAL_STATE);
    }

    private byte[] hasher;

    void prepareNativeState() {
        // effectively, it is `blake3_hasher_init(hasher)`. We cache the result to INITIAL_STATE and do a memcpy to avoid JNI roundtrip.
        System.arraycopy(INITIAL_STATE, 0, hasher, 0, STATE_SIZE);
    }

    void callNativeMethodThatNeedAHasher() {
        // Similar to previous checks we had: `hasher == -1`
        if (needPrepareNativeState) {
            prepareNativeState();
        }
        nativeMethodThatNeedAHasher(hasher);
    }

    private static native void nativeInit(byte[] hasher);
    private static native void nativeMethodThatNeedAHasher(byte[] hasher);
}

// JNI
... void xxx_nativeInit(..., jbyteArray jhasher) {
    blake3_hasher *hasher = (blake3_hasher *)env->GetPrimitiveArrayCritical(jstate, nullptr);
    if (hasher) {
        blake3_hasher_init(hasher);
        env->ReleasePrimitiveArrayCritical(jhasher, hasher, JNI_ABORT);
    }
}

... void xxx_nativeMethodThatNeedAHasher(..., jbyteArray jhasher) {
    blake3_hasher *hasher = (blake3_hasher *)env->GetPrimitiveArrayCritical(jstate, nullptr);
    if (hasher) {
        // Do anything with hasher.
        env->ReleasePrimitiveArrayCritical(jhasher, hasher, JNI_ABORT);
    }
}
```

Note that, when passing byte array to JNI, we just pass a pointer, not copying the entire array. We use `GetPrimitiveArrayCritical` to get the underlying buffer which may do a copy, but most of the time there is no copy.

Reply:
@couevre As far as I understood the code, `close` is an optional call here, which is a common pattern with `Cleaner`: Immediate release of resources if the caller needs it by calling `close`, guaranteed automated release eventually via the `Cleaner`.

If Blaze successfully uses `GetPrimitiveArrayCritical` internally, it's probably better to use the same pattern here to reduce the risk of performance regressions that go unnoticed because the Bazel code path isn't covered by google3.

Reply:
@coeuvre Thanks for the example, this is perfect.

Done.

Reply:
Seems like there's now a linking issue due to us doing JNI stuff in static {} -- any ideas on how / why this is happening? I consolidated the JNI code into Blake3MessageDigest so that I could ensure it was running in order and JNI was being loaded before use, but still seeing this.

https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/01896762-aba0-4951-9be6-b3568825b687/src/test/shell/bazel/bazel_bootstrap_distfile_tar_test/test.log

Reply:
That usage is fine as it comes after the `loadJni` call. The failing test bootstraps Bazel and may not honor BUILD files, maybe the new JNI sources need to be listed somewhere else?

Reply:
@fmeum -- that was my intuition too but I can't figure out where to add these. `blake3_jni.cc` should already be a part of the `src/main/native:srcs` filegroup afaict?

```
filegroup(
    name = "srcs",
    srcs = glob(["**"]) + [
        "//src/main/native/windows:srcs",
    ],
    visibility = ["//src:__pkg__"],
)
```

Reply:
I think you need to trigger JNI loader similar to https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/unix/NativePosixSystem.java;l=28;drc=50adefb987351eb3612b2e28240cf0d0697fdf5a.

Reply:
Oops, nevermind, you already did that. Let me dig more.

Reply:
Okay, I found the reason. For the bootstrap test, we compile stage1 Bazel using [compile.sh](https://cs.opensource.google/bazel/bazel/+/master:scripts/bootstrap/compile.sh;l=408;drc=30b21303b0fd65c70285ea0832db18f78ad9fd0f) where we don't build JNI but fallback to use NIO. We build a stage2 Bazel using the binary from stage1, which doesn't have the JNI, so we failed to initialize Blake3MessageDigest.

How it works for `UnixFileSystem` is, we [detect whether JNI is available](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/BazelFileSystemModule.java;l=75;drc=6d6bbdcd7ac7e8130741417d891e641495b046b2). If not fallback to use NIO based file system.

In order to pass the test, we should do something similar in `BazelHashFunctions` where we only register Blake3 if JNI is available.

Reply:
Done.